sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala

```diff
@@ -70,4 +74,69 @@ case class ShuffledHashJoinExec(
       join(streamIter, hashed, numOutputRows)
     }
   }
+
+  override def inputRDDs(): Seq[RDD[InternalRow]] = {
+    streamedPlan.execute() :: buildPlan.execute() :: Nil
+  }
+
+  override def needCopyResult: Boolean = true
+
+  override protected def doProduce(ctx: CodegenContext): String = {
+    // inline mutable state since not many join operations in a task
+    val streamedInput = ctx.addMutableState(
+      "scala.collection.Iterator", "streamedInput", v => s"$v = inputs[0];", forceInline = true)
+    val buildInput = ctx.addMutableState(
+      "scala.collection.Iterator", "buildInput", v => s"$v = inputs[1];", forceInline = true)
+    val initRelation = ctx.addMutableState(
+      CodeGenerator.JAVA_BOOLEAN, "initRelation", v => s"$v = false;", forceInline = true)
+    val streamedRow = ctx.addMutableState(
+      "InternalRow", "streamedRow", forceInline = true)
+
+    val thisPlan = ctx.addReferenceObj("plan", this)
+    val (relationTerm, _) = prepareRelation(ctx)
+    val buildRelation = s"$relationTerm = $thisPlan.buildHashedRelation($buildInput);"
+    val (streamInputVar, streamInputVarDecl) = createVars(ctx, streamedRow, streamedPlan.output)
+
+    val join = joinType match {
+      case _: InnerLike => codegenInner(ctx, streamInputVar)
+      case LeftOuter | RightOuter => codegenOuter(ctx, streamInputVar)
+      case LeftSemi => codegenSemi(ctx, streamInputVar)
+      case LeftAnti => codegenAnti(ctx, streamInputVar)
+      case _: ExistenceJoin => codegenExistence(ctx, streamInputVar)
+      case x =>
+        throw new IllegalArgumentException(
+          s"ShuffledHashJoin should not take $x as the JoinType")
+    }
+
+    s"""
+       |// construct hash map for shuffled hash join build side
+       |if (!$initRelation) {
+       |  $buildRelation
+       |  $initRelation = true;
+       |}
+       |
+       |while ($streamedInput.hasNext()) {
+       |  $streamedRow = (InternalRow) $streamedInput.next();
+       |  ${streamInputVarDecl.mkString("\n")}
+       |  $join
+       |
+       |  if (shouldStop()) return;
+       |}
+     """.stripMargin
+  }
+
+  /**
+   * Returns a tuple of variable name for HashedRelation,
+   * and boolean false to indicate key not to be known unique in code-gen time.
+   */
+  protected override def prepareRelation(ctx: CodegenContext): (String, Boolean) = {
+    if (relationTerm == null) {
+      // Inline mutable state since not many join operations in a task
+      relationTerm = ctx.addMutableState(
+        "org.apache.spark.sql.execution.joins.HashedRelation", "relation", forceInline = true)
```

Comment:
As you already use mutable state for the hashed relation here, why don't just follow BroadcastHashJoinExec to call buildHashedRelation at prepareRelation and set it to mutable state? Then BroadcastHashJoinExec and ShuffledHashJoinExec look more consistent.

Reply:
@viirya - `BroadcastHashJoinExec` needs to [broadcast build side](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala#L222-L223) in `prepareBroadcast()`. I feel it's hard to refactor there. Wondering do you have any idea to make it cleaner?

Reply:
I have the same feeling with @viirya .

Can we broadcast build side in `BroadcastHashJoinExec.prepareRelation`?

Reply:
We can always call `prepareRelation` in `doProduce` as the first step
```
  override def doProduce(ctx: CodegenContext): String = {
    relationTerm = prepareHashedRelation(ctx)
    streamedPlan.asInstanceOf[CodegenSupport].produce(ctx, this)
  }
```

And declare `useUniqueKeyCodePath` as a separated method.

Reply:
in BroadcastHashJoinExec
```
  // Note that this will be called in `doConsume`, and the broadcast is already finished in
  // `doProduce`.
  @transient
  protected lazy val useUniqueKeyCodePath: Boolean = {
    require(broadcastRelation != null,
      "Don't evalulate `useUniqueKeyCodePath` lazy val until `broadcastRelation` is assigned.")
    broadcastRelation.value.keyIsUnique
  }
```

Reply:
Sorry I may not understand the real problem. What's wrong with
```
  override def doProduce(ctx: CodegenContext): String = {
    streamedPlan.asInstanceOf[CodegenSupport].produce(ctx, this)
  }

  override def doConsume(ctx: CodegenContext, input: Seq[ExprCode], row: ExprCode): String = {
    joinType match {
      case _: InnerLike => codegenInner(ctx, input)
      case LeftOuter | RightOuter => codegenOuter(ctx, input)
      case LeftSemi => codegenSemi(ctx, input)
      case LeftAnti => codegenAnti(ctx, input)
      case j: ExistenceJoin => codegenExistence(ctx, input)
      case x =>
        throw new IllegalArgumentException(
          s"BroadcastHashJoin should not take $x as the JoinType")
    }
  }
```

Reply:
@cloud-fan and @viirya -

Updated code with following change:
 
(1). `ShuffledHashJoinExec.prepareRelation` will do `buildRelation` to build hash map.
(1).`BroadcastHashJoinExec.prepareRelation` will do `prepareBroadcast`to broadcast build side and build hash map.

Reply:
@cloud-fan 

re https://github.com/apache/spark/pull/29277#discussion_r462449882:

wondering do you think keeping two vars would look much better, than the current approach for calling `prepareRelation` separately in each `codegenInner/codegenOuter/...` ? If yes, I can make the change accordingly, thanks.

re https://github.com/apache/spark/pull/29277#discussion_r462457807:

By design, `doConsume()` generates code for processing one input row. `BroadcastHashJoinExec` can do codegen work in `doConsume()` with only stream side input, because it just broadcasts executing its build side query plan, and generates per-row-processing codegen for stream side in `doConsume()`.

However, `ShuffledHashJoinExec` cannot do codegen work in `doConsume()` with stream and build side input, because it needs to first read all build side input rows and build a hash map, before processing each row from stream side input. We cannot generate code in `doConsume()` with simply a pair of stream and build side input row. Similar to `SortMergeJoinExec`, where it needs to stream only one row in one side, and buffer the other side for multiple rows. `ShuffledHashJoinExec` has to generate code in `doProduce()`, and its children have to do codegen separately in their own iterator classes.